### PR TITLE
Add create command to kidlectl (#21)

### LINKED ISSUE #21: add a kidectl create command to create an IdlingResource
Add a new command `create` on `kidlectl`. This command creates an `IdlingResource` in a fancy way.
The CLI flags has to be defined.

- [ ] add the core create command
- [ ] add the flags for the cronjob strategy

### PR DESCRIPTION
**Pull request by doniacld**
### Description
Add create idling resource command to kidlectl to close #21.

### Test

I had to comment the installation of kustomize to make the installation works.
I could test manually:

* Creation of a resource without passing the idle flag
```
[donia@localhost kidlectl]$ ./bin/kidlectl create kidle-test --kubeconfig=~/.kubeconfig -n default -r deployments/kidle
2021-10-26T20:57:03.186+0200	INFO	creating the idling resource	{"namespace": "default", "name": "kidle-test", "ref": "deployments/kidle"}
2021-10-26T20:57:03.204+0200	INFO	idling resource created	{"namespace": "default", "name": "kidle-test", "ref": "deployments/kidle"}
``` 

```
[donia@localhost ~]$ kubectl get idlingresources
NAME         IDLE    REFKIND       REFNAME
kidle-test   false   deployments   kidle
```
* Creation of a resource with the idle flag
```
[donia@localhost kidlectl]$ ./bin/kidlectl create kidle-test --kubeconfig=~/.kubeconfig -n default -r deployments/kidle -i true
2021-10-26T21:04:47.087+0200	INFO	creating the idling resource	{"namespace": "default", "name": "kidle-test", "ref": "deployments/kidle"}
2021-10-26T21:04:47.105+0200	INFO	idling resource created	{"namespace": "default", "name": "kidle-test", "ref": "deployments/kidle"}
``` 

```
[donia@localhost ~]$ kubectl get idlingresources
NAME         IDLE    REFKIND       REFNAME
kidle-test   true    deployments   kidle
```

* Attempt of creation with an unknown kind
```
[donia@localhost kidlectl]$ ./bin/kidlectl create kidle-test --kubeconfig=~/.kubeconfig -n default -r foo/kidle -i true
2021-10-28T19:33:19.388+0200	INFO	creating the idling resource	{"namespace": "default", "name": "kidle-test", "ref": "foo/kidle"}
2021-10-28T19:33:19.472+0200	ERROR	unable to create an idling resource	{"error": "invalid resource kind; got `foo` expected one of: cj, cronjob, cronjobs, deploy, deployment, deployments, statefulset, statefulsets, sts"}
``` 

### TODO

- [x] Add the validation kind resource call
- [x] Add the `kidlectl create` command to the manifest in the userguide

### CODE REVIEW

File: cmd/kidlectl/pkg/kidleclient.go [hunk 1]
```diff
@@ -80,3 +86,42 @@ func (k *KidleClient) applyDesiredIdleState(idle bool, req *client.ObjectKey) (b
...
+		Spec: kidlev1beta1.IdlingResourceSpec{
+			IdlingResourceRef: kidlev1beta1.CrossVersionObjectReference{
+				Kind:       values[0],
+				Name:       values[1],
+				APIVersion: "apps/v1",
+			},
+			Idle: idle,
+		},
+	}
+	logf.Log.Info("Idling resource", "ir", ir)
```
Comment (doniacld) [inline cmd/kidlectl/pkg/kidleclient.go]: remove this line

File: cmd/kidlectl/pkg/kidleclient.go [hunk 2]
```diff
@@ -80,3 +86,42 @@ func (k *KidleClient) applyDesiredIdleState(idle bool, req *client.ObjectKey) (b
...
+// CreateIdlingResource creates an IdlingResource with the given values
+func (k *KidleClient) CreateIdlingResource(idle bool, ref string, req *client.ObjectKey) (bool, error) {
+	ctx := context.Background()
+	logf.Log.Info("entering idling resource")
+
+	values := strings.Split(ref, "/")
+	if len(values) < 2 {
+		return false, fmt.Errorf("invalid idlingresource ref; expected <kind>/<name> got %s", ref)
+	}
+	// TODO validate if the kind is known
```
Comment (orphaner) [inline cmd/kidlectl/pkg/kidleclient.go]: please have a look at this commit: https://github.com/kidle-dev/kidle/commit/81c280bc325182f9f7cfe9674d916723275b1444

**Comment by orphaner** [comment]
hello @doniacld; how can we get organized so that we can work together finishing this PR?

**Comment by doniacld** [comment]
> hello @doniacld; how can we get organized so that we can work together finishing this PR?

Hey @orphaner!
Sorry for the delay, I pinged you on Slack yesterday.
I have been able to test the written code 🍾 
Everything seems to work, I added the code from the discovery in the refactoring.

I just have one problem from Sonar, it points a duplication between [idle.go](https://github.com/kidle-dev/kidle/blob/d206ccbfa51f5ee2698432548547fa306be58a84/cmd/kidlectl/idle.go) and [wakeup.go](https://github.com/kidle-dev/kidle/blob/d206ccbfa51f5ee2698432548547fa306be58a84/cmd/kidlectl/wakeup.go) even if the logs are different.
If you have any hint about it, I'll take it to pass the CI. 
I would be happy yo have your review and discuss the code.

File: cmd/kidlectl/main.go [hunk 3]
```diff
@@ -1,22 +1,21 @@
...
-	IdleCmd    idleCommandOptions    `command:"idle" alias:"i" description:"idle the referenced object of an IdlingResource"`
-	WakeUpCmd  wakeupCommandOptions  `command:"wakeup" alias:"w" description:"wakeup the referenced object of an IdlingResource"`
-	VersionCmd versionCommandOptions `command:"version" description:"show the kidle version information"`
+// Options are the cli main options for go-flags
+type Options struct {
+	Kubeconfig   string                  `long:"kubeconfig" env:"KUBECONFIG" description:"path to Kubernetes config file"`
+	IdleCmd      idleCommandOptions      `command:"idle" alias:"i" description:"idle the referenced object of an IdlingResource"`
+	WakeUpCmd    wakeupCommandOptions    `command:"wakeup" alias:"w" description:"wakeup the referenced object of an IdlingResource"`
+	CreateCmd    createCommandOptions    `command:"create" alias:"c" description:"create an IdlingResource"`
+	DiscoveryCmd discoveryCommandOptions `command:"discovery" description:"discover allowed Group Version Kind"`
```
Comment (orphaner) [inline cmd/kidlectl/main.go]: to remove ; on my branch it was part of the proof of concept; there is no need to keep it here

File: cmd/kidlectl/main.go [hunk 4]
```diff
@@ -59,57 +72,14 @@ func main() {
...
-		if done {
-			logf.Log.V(0).Info("waked up", "namespace", kidle.Namespace, "name", opts.WakeUpCmd.Args.Name)
-		} else {
-			logf.Log.V(0).Info("already waked up", "namespace", kidle.Namespace, "name", opts.WakeUpCmd.Args.Name)
-		}
-
+		cmdWakeup(opts.WakeUpCmd)
+	case "create":
+		cmdCreate(opts.CreateCmd)
+	case "discovery":
```
Comment (orphaner) [inline cmd/kidlectl/main.go]: to remove

**Review by orphaner — CHANGES_REQUESTED** [review]

File: cmd/kidlectl/main.go (hunk 4 above)
Comment (doniacld) [reply cmd/kidlectl/main.go]: Fixed.

File: cmd/kidlectl/main.go (hunk 3 above)
Comment (doniacld) [reply cmd/kidlectl/main.go]: Fixed, I removed the discovery command.

**Comment by orphaner** [comment]
LGTM ; it's now merged :tada:

**Pull request merged**